Scala3 Documentation (#3)
scalacenter/docs.scala-lang: _overviews/scala3-book/scala-for-javascript-devs.md

  padding-top: 0.6em;
}
.content-primary table th {
  text-align: center;
}
</style>

This page provides a comparison between the JavaScript and Scala programming languages. It’s intended for programmers who know JavaScript and want to learn about Scala, specifically by seeing examples of how JavaScript language features compare to Scala.

## Overview 

This section provides a relatively brief introduction and summary of the sections that follow. It presents the similarities and differences between JavaScript and Scala at a high level, and then introduces the differences you’ll experience every day as you write code.

### High-level similarities 

- At a high level, Scala shares these similarities with JavaScript:  
- Both are considered high-level programming languages, where you don’t have to concern yourself with low-level concepts like pointers and manual memory management
- Both have a relatively simple, concise syntax
- Both support a C/C++/Java style curly-brace syntax for writing methods and other block of code  
- Both are object-oriented programming (OOP) languages
- Both are functional programming (FP) languages, with support for lambdas and higher-order functions

>   - Functions are first-class citizens in both languages  

Author: I agree with you that language with support for lambdas is not to be equated with FP. _However_, while I also am in strong favor of immutability, I (and probably large parts of the Racket / Scheme community) would disagree that FP is about immutability first :)

With this off my chest, I propose we simply drop the FP in this comparison.

> - JavaScript runs in the browser, and thanks to [the *Scala.js* project](https://www.scala-js.org/), Scala can be compiled to JavaScript to run in the browser

Author: Thanks, I'll adapt your suggestion.

  val numLegs = 4
  def walk() = println("I’m walking")
  def wagTail() = println("⎞⎜⎛  ⎞⎜⎛")
  def stopTail() = println("Tail is stopped")
  override def toString = s"$name is a Dog"

// create a Dog instance
val d = Dog("Rover")

// use the class’s attributes and behaviors
println(d.numLegs)   // 4
d.wagTail()          // "⎞⎜⎛  ⎞⎜⎛"
d.walk()             // "I’m walking"
```

## Control Structures 

Except for the use of `===` and `!==` in JavaScript, comparison and logical operators are almost identical in JavaScript and Scala:  

<table>
  <tbody>
    <tr>
      <th valign="top">JavaScript</th>
      <th valign="top">Scala</th>
    </tr>
    <tr>
      <td rowspan="1" colspan="2" class="table-desc-row">Comparison operators</td>
    </tr>
    <tr>
      <td valign="top"><code>==</code></td>
      <td valign="top"><code>==</code></td>
    </tr>
    <tr>
      <td valign="top"><code>===</code></td>
      <td valign="top"><code>==</code></td>
    </tr>
    <tr>
      <td valign="top"><code>!=</code></td>
      <td valign="top"><code>!=</code></td>
    </tr>
    <tr>
      <td valign="top"><code>!==</code></td>
      <td valign="top"><code>!=</code></td>

>     </tr>

Author: I suggested to @alvinj to list them not next to each other but one after another (below each other), this would also would make it less likely to suggest that they are equivalent.

>     </tr>

Reviewer: We are going to reformat this page (and the Python and Java) pages to show one example after the other. Is there any other text I should add here? I haven’t used plain JavaScript in a long time, and I didn’t know they are so dissimilar.

        <br>do
        <br>&nbsp; val i = i * 2
        <br>&nbsp; println(j)
        <br>
        <br>// also available
        <br>for (i &lt;- nums) {
        <br>&nbsp; val j = i * 2
        <br>&nbsp; println(j)
        <br>}</code>
      </td>
    </tr>
    <tr>
      <td rowspan="1" colspan="2" class="table-desc-row">Breaking a <code>for</code>
        loop</td>
    </tr>
    <tr>
      <td valign="top"><code>for (let i=0; i&lt;10; i++) {
        <br>&nbsp; if (i &gt; 5) break;
        <br>&nbsp; console.log(i);
        <br>}</code>
      </td>
      <td valign="top">A <code>break</code> statement is very rarely
        used in Scala. If you need it, you can import it and then
        use it.
      </td>
    </tr>
    <tr>
      <td rowspan="1" colspan="2" class="table-desc-row">Multiple generators</td>
    </tr>
    <tr>
      <td valign="top"><code>for (let i = 0; i &lt; 2; i++) {
        <br>&nbsp; for (let j = 0; j &lt; 3; j++) {
        <br>&nbsp;&nbsp;&nbsp; console.log(`i: ${i} j: ${j}`);
        <br>&nbsp; }
        <br>}</code>
      </td>
      <td valign="top"><code>for
        <br>&nbsp; i &lt;- 1 to 2
        <br>&nbsp; j &lt;- 'a' to 'b'

>         <br>&nbsp; k &lt;- 1 to 10 by 5

Reviewer: I updated this example, but it doesn’t fit with the current formatting, so I commented it out until the page is reformatted.

      <td valign="top"><code>for (let i = 0; i &lt; 10; i++) {
        <br>&nbsp; if (i % 2 == 0 &amp;&amp; i &lt; 5) {
        <br>&nbsp;&nbsp;&nbsp; console.log(i);
        <br>&nbsp; }
        <br>}</code>
      </td>
      <td valign="top"><code>for
        <br>&nbsp; i &lt;- 1 to 10
        <br>&nbsp; if i % 2 == 0
        <br>&nbsp; if i &lt; 5
        <br>do
        <br>&nbsp; println(i)</code>
      </td>
    </tr>
    <tr>
      <td rowspan="1" colspan="2" class="table-desc-row"><code>for</code>
        comprehension</td>
    </tr>
    <tr>
      <td valign="top">N/A</td>
      <td valign="top"><code>// a for-comprehension is a for-loop
        <br>// that returns/yields a value
        <br>val list = 
        <br>&nbsp; for
        <br>&nbsp;&nbsp;&nbsp; i &lt;- 1 to 3
        <br>&nbsp; yield
        <br>&nbsp;&nbsp;&nbsp; i * 10
        <br>// result: Vector(10, 20, 30)</code>
      </td>
    </tr>
  </tbody>
</table>

### switch & match 

Where JavaScript has `switch` statements, Scala has `match` expressions. Like everything else in Scala, these truly are *expressions*, meaning they return a result.

```scala
// a sample match expression that returns a result

> var day = 1

Reviewer 2: ```suggestion
val day = 1
```

```scala
d(0)   // "eleven"
d(1)   // 11
```

### Enumerations

JavaScript doesn’t have enumerations, but you can do this:

```javascript
let Color = {
  RED: 1,
  GREEN: 2,
  BLUE: 3
};
Object.freeze(Color);
```

In Scala 3 you can do quite a few things with enumerations. You can create an equivalent of that code:

```scala
enum Color:
  case Red, Green, Blue
```

You can create a parameterized enum:  

```scala
enum Color(val rgb: Int):
  case Red   extends Color(0xFF0000)
  case Green extends Color(0x00FF00)
  case Blue  extends Color(0x0000FF)
```

You can also create user-defined enum members:

```scala
enum Planet(mass: Double, radius: Double):
  case Mercury extends Planet(3.303e+23, 2.4397e6)
  case Venus   extends Planet(4.869e+24,6.0518e6)
  case Earth   extends Planet(5.976e+24,6.37814e6)
  // more planets here ...

  private final val G = 6.67300E-11
  def surfaceGravity = G * mass / (radius * radius)
  def surfaceWeight(otherMass: Double) = otherMass * surfaceGravity
```

## Scala.js DOM Code

> Scala.js is a project that lets you write Scala code that is compiled to JavaScript code that can then be used in the browser. The approach is similar to CoffeeScript, TypeScript, and other languages that are compiled (“transpiled”) to JavaScript.  

Reviewer 2: ```suggestion
Scala.js lets you write Scala code that is compiled to JavaScript code that can then be used in the browser. The approach is similar to TypeScript, ReScript, and other languages that are compiled to JavaScript.  
```
(please don't say "transpiled"; this word means nothing)

Author: haha :) true

Once you include the necessary libraries, and import the necessary packages in your project, writing Scala.js code looks very similar to writing JavaScript code:  

```scala
// show an alert dialog on a button click

> $("#hello-button").click{() =>

Reviewer 2: ```suggestion
jQuery("#hello-button").click{() =>
```
`$` is reserved in Scala.

  dom.window.alert("Hello, world")
}

// define a button and what should happen when it’s clicked
val btn = button(
  "Click me",

>   onclick := { () =>
>     dom.window.alert("Hello, world")
> })

Reviewer 2: ```suggestion
  onclick := { () =>
    dom.window.alert("Hello, world")
  })
```

// create two divs with css classes, an h2 element, and the button
val content =

>   div(cls := "foo",
>     div(cls := "bar",
>       h2("Hello"),
>         btn
>       )
>   )

Reviewer 2: ```suggestion
  div(cls := "foo",
    div(cls := "bar",
      h2("Hello"),
      btn
    )
  )
```

// add the content to the DOM
val root = dom.document.getElementById("root")
root.innerHTML = ""
root.appendChild(content.render)
```

> Notice that although Scala is a type-safe language, no types are declared in that code. Scala’s strong type inference capabilites often make Scala code look like it’s dynamically-typed. But it is type-safe, so you catch many classes of errors early in the development cycle.  

Reviewer 2: ```suggestion
Note that although Scala is a type-safe language, no types are declared in that code. Scala’s strong type inference capabilities often make Scala code look like it’s dynamically typed. But it is type-safe, so you catch many classes of errors early in the development cycle.  
```

## Other Scala.js resources

> The Scala.js website has an excellent collection of tutorials from JavaScript developers interested in using Scala.js. Here are some of their initial tutorials:  

Reviewer 2: ```suggestion
The Scala.js website has an excellent collection of tutorials for JavaScript developers interested in using Scala.js. Here are some of their initial tutorials:  
```

- [Basic tutorial (creating a first Scala.js project)](https://www.scala-js.org/doc/tutorial/basic/)  
- [Scala.js for JavaScript developers](https://www.scala-js.org/doc/sjs-for-js/)

> - [From ES6 to Scala: Basics  
>     ](https://www.scala-js.org/doc/sjs-for-js/es6-to-scala-part1.html)

Reviewer 2: ```suggestion
- [From ES6 to Scala: Basics](https://www.scala-js.org/doc/sjs-for-js/es6-to-scala-part1.html)
```

- [From ES6 to Scala: Collections](https://www.scala-js.org/doc/sjs-for-js/es6-to-scala-part2.html)
- [From ES6 to Scala: Advanced](https://www.scala-js.org/doc/sjs-for-js/es6-to-scala-part3.html)  

## Concepts that are unique to Scala 

There are other concepts in Scala which currently have no equivalent in JavaScript:  

- Almost everything related to contextual abstractions
- Method features:
  -   Multiple parameter lists

>   -   Default parameter values

Reviewer 2: This also exists in JavaScript.

Author: Just to get you right, you are _only_ talking about default params, right?
